### ISSUE TITLE
MCMC: how to support a target_log_prob_fn with a random seed?

### ISSUE THREAD

**Issue by berthyf96** [tensorflow/probability#1747]
I'm using TFP on JAX to perform MCMC sampling. I have a `target_log_prob_fn` that has randomness to it because it computes an unbiased estimate of the log-prob. So my log-probs can only be computed as `log_prob = target_log_prob_fn(rng, state)`, and I would like to be able to split the seed at each MCMC step.

How can I use the TFP API with this type of `target_log_prob_fn`?

**Comment by brianwa84** [comment]
I think you could write a TransitionKernel subclass whose one_step (which
receives a seed arg) creates an appropriate builtin kernel and calls it's
one_step with a functools.partial over the stochastic logprob.

You would need to be careful with hmc because it keeps a gradient in the
state, which would need replacement. The implementation of ReplicaExchange
would be informative on that point.

**Comment by ColCarroll** [comment]
You might also try out [blackjax](https://github.com/blackjax-devs/blackjax) to build a sampler -- the components there are pretty easy to swap in and out, and work fine with TFP-on-JAX densities. Specifically, I will often copy/paste functions that _almost_ work, get them to work for my use case, then pass them to other parts of blackjax so I don't have to, e.g., write my own adaptation code.

**Comment by berthyf96** [comment]
Thank you! I think I got a decent hack working by subclassing TransitionKernel. Will look into blackjax if I run into issues with this.

```
import functools

import jax
import tensorflow_probability.substrates.jax as tfp

from tensorflow_probability.python.mcmc import kernel as kernel_base


class StochasticLogProb(kernel_base.TransitionKernel):
  def __init__(self, target_log_prob_fn, make_kernel_fn, name=None):
    self._parameters = {k: v for k, v in locals().items() if v is not self}

  @property
  def target_log_prob_fn(self):
    return self._parameters['target_log_prob_fn']

  @property
  def make_kernel_fn(self):
    return self._parameters['make_kernel_fn']

  @property
  def name(self):
    return self._parameters['name']

  @property
  def is_calibrated(self):
    return True

  def bootstrap_results(self, init_state, seed=jax.random.PRNGKey(1)):
    seed = tfp.random.sanitize_seed(seed)
    target_log_prob_for_inner_kernel = functools.partial(self.target_log_prob_fn, seed=seed)
    inner_kernel = self.make_kernel_fn(target_log_prob_for_inner_kernel)
    results = inner_kernel.bootstrap_results(init_state)
    return results
  
  def one_step(self, current_state, previous_kernel_results, seed=None):
    seed = tfp.random.sanitize_seed(seed)
    target_log_prob_seed, inner_seed = tfp.random.split_seed(seed, n=2)
    
    target_log_prob_for_inner_kernel = functools.partial(self.target_log_prob_fn, seed=target_log_prob_seed)
    inner_kernel = self.make_kernel_fn(target_log_prob_for_inner_kernel)

    # Run a step of the inner TransitionKernel.
    state, results = inner_kernel.one_step(current_state, previous_kernel_results, seed=inner_seed)
    return state, results
```
